paddle/fluid/inference/analysis/passes/memory_optimize_pass.cc

```diff
@@ -0,0 +1,647 @@
+    item.second /= num_batch;
+  }
+
+  return var2size;
+}
+
+// Analysis the batch shapes loading from the cache file.
+// By splitting the variables to different clusters by analyzing their batch
+// size, we can pre-schedule the changes of difference LoDTensor when different
+// length of input sequences is entered.
+// This should works fine for the models operating on sentences.
+std::vector<std::unordered_set<std::string>> AnalysisBatchShapesByBatchSize(
+    const std::vector<std::map<std::string, std::vector<int>>>& batches) {
+  // collect the batch size of each shape and combine to a stringstream in
+  // converient to generate a hash.
+  std::unordered_map<std::string, std::stringstream> var_batchsize_hashes;
+  for (auto& batch : batches) {
+    for (auto& ele : batch) {
+      int batch_size = ele.second.front();
+      // TODO(Superjomn) might consume large memory here, use combine hash.
+      var_batchsize_hashes[ele.first] << batch_size;
+    }
+  }
+
+  // Split to sets by batch size sequences.
+  std::unordered_map<size_t /*hash*/, std::unordered_set<std::string>>
+      shape_sets;
+  for (auto& ele : var_batchsize_hashes) {
+    auto hash = std::hash<std::string>()(ele.second.str());
+    shape_sets[hash].insert(ele.first);
+  }
+  std::vector<std::unordered_set<std::string>> res;
+  for (auto& ele : shape_sets) {
+    res.emplace_back(std::move(ele.second));
+  }
+
+  VLOG(3) << "Cluster by batch_size and get " << res.size() << " clusters";
+  return res;
+}
+
+// Analysis the batch shapes loading from the cache file, and split them to
+// different clusters by their size.
+// This should works fine for the overall models.
+std::vector<std::unordered_set<std::string>> AnalysisBatchShapesBySimilarSize(
+    const space_table_t& space_table,
+    const std::vector<std::map<std::string, std::vector<int>>>& batches,
+    int interval = 200000) {
+  PADDLE_ENFORCE_GT(interval, 0);
+  // cluster to different clusters.
+  size_t max_size = 0;
+  for (auto& item : space_table) {
+    max_size = std::max(item.second, max_size);
+  }
+  VLOG(4) << "tensor max size " << max_size;
+
+  std::vector<std::unordered_set<std::string>> res;
+
+  // cluster by intervals.
+  for (size_t interval_size = 0; interval_size <= max_size;
+       interval_size += interval) {
+    std::unordered_set<std::string> cluster;
+    for (auto& item : space_table) {
+      if (interval_size <= item.second &&
+          interval_size + interval > item.second) {
+        cluster.insert(item.first);
+      }
+    }
+    if (!cluster.empty()) {
+      res.push_back(cluster);
+    }
+  }
+
+  VLOG(3) << "Cluster by interval and get " << res.size() << " cluster";
+  return res;
+}
+
+std::string MemoryOptimizePass::repr() const { return "memory optimize pass"; }
+
+void MemoryOptimizePass::RunImpl(Argument* argument) {
+  // When force update, should not optimize memory.
+  if (!argument->enable_memory_optim() || argument->memory_optim_force_update())
```

Comment:
this should normally live in pass manager

Reply:
Currently, the inference memory optimize performs in two stages:

1. Execute the inference program on the real data, store the shapes of variables into the disk. This stage is offline.
2. Online in the real deployment, load the shapes persisted on the disk, perform the memory reusing strategy, and run inference. This phase is where the memory optimize performs.

The `enable_memory_optim` config will make the predictor perform those two stages, where the first stage is not suitable to perform in the real online deployment, so the config needs to be carefully set in the real application, and cannot be switched on by default.

We will add another simple strategy that doesn't need the first stage, and then, the memory optimize can be turned on by default.